### PR TITLE
Extend DateOnly/TimeOnly support to include DataTable as structured parameter

### LINKED ISSUE #2209: Limited DateOnly C# support when using SqlParameter.Structured
### C# DateOnly not supported when using SqlParameter.Structured (Datatable)
Current .net 8 rc2 works with DateOnly type when passing DateOnly values to sql procedure like so:
image

But if we try to pass the DateOnly as udtt via SqlParameter.Structured we get an error like so:
image

#### Program.cs

```c#
 internal class Program
 {
     static async Task Main(string[] args)
     {
         var x = new TestAppSimple();
         await x.TestDataGet(); 
     }
 }
```

#### TestAppSimple.cs

```c#
 internal class TestAppSimple
{
    private SqlConnection Connection { get; set; }

    public TestAppSimple()
    {
        var connectionString = "Server=localhost;..."; 
        Connection = new SqlConnection(connectionString);

    }

    public async Task TestDataGet()
    {
        try
        {
            Connection.Open();

            await TryToPassDateOnlyAsParamSimple(); //Works!
            await TryToPassDateOnlyAsStructuredParamSimple(); //Fails!
        }
        catch(Exception e)
        {
            Console.WriteLine(e.Message);
        }
        finally
        {
            Connection?.Close();
            Connection?.Dispose();
        }
    }

    #region Sql tests

    private async Task TryToPassDateOnlyAsParamSimple()
    {
        try
        {
            var command = Connection.CreateCommand();
            command.CommandText = "[dbo].[spTestDateOnly]";
            command.CommandTimeout = 30;
            command.CommandType = CommandType.StoredProcedure;
            command.Parameters.Add(new SqlParameter
            {
                ParameterName = "@date",
                SqlDbType = SqlDbType.Date,
                Value = new DateOnly(2023, 11, 15)
            });

            using (var reader = await command.ExecuteReaderAsync())
            {
                //if passes here, then it worked.
            }

        }
        catch (Exception e)
        {
            Console.WriteLine(e.Message);
        }
    }

    private async Task TryToPassDateOnlyAsStructuredParamSimple()
    {
        try
        {
            var command = Connection.CreateCommand();
            command.CommandText = "[dbo].[spTestDateOnlyUdtt]";
            command.CommandTimeout = 30;
            command.CommandType = CommandType.StoredProcedure;

            //init param data
            var dates = new List<DateOnly>(new[] { new DateOnly(2023, 11, 15), new DateOnly(2022, 10, 20) });

            //populate dt parameter
            var dtDates = new DataTable();
            dtDates.Columns.Add(new DataColumn("Date", typeof(DateOnly)));

            foreach (var date in dates)
            {
                var dataRow = dtDates.NewRow();
                dataRow["Date"] = date;
                dtDates.Rows.Add(dataRow);
            }

            //pass dt parameter to procedure
            command.Parameters.Add(new SqlParameter
            {
                ParameterName = "@dates",
                SqlDbType = SqlDbType.Structured,
                TypeName = "[dbo].[udttDateOnly]",
                Value = dtDates
            });


            using (var reader = await command.ExecuteReaderAsync())
            {
                //if passes here, then it worked.
            }
        }
        catch (Exception e)
        {
            Console.WriteLine(e.Message);
        }

    }

    #endregion
   
}
```

#### sql - udttDateOnly
```sql
  CREATE TYPE [dbo].[udttDateOnly] AS TABLE(
     [Date] DATE NULL
 )
 GO
```

#### sql - spTestDateOnlyUdtt
```sql
 CREATE OR ALTER PROC [dbo].[spTestDateOnlyUdtt]
 (
   @dates [dbo].[udttDateOnly] READONLY
 )
 AS
     SET NOCOUNT ON;

     DECLARE @context NVARCHAR(255) = '[spTestDateOnlyUdtt]'

     SELECT COUNT(*) FROM @dates;
	
 GO

```

#### sql - spTestDateOnly
```sql
CREATE OR ALTER PROC [dbo].[spTestDateOnly]
(
    @date DATE
)
AS
    SET NOCOUNT ON;

    DECLARE @context NVARCHAR(255) = '[spTestDateOnly]'

    PRINT @date
	
GO

```

### Expected behavior
SqlParameter.Structured should pass the provided/populated udtt to the sql procedure.

### Further technical details
Microsoft.Data.SqlClient version: 5.2.0-preview3.23201.1
.NET target: 8.0.100-rc.2.23502.2
SQL Server version:  SQL Server 16.0.1105.1
Operating system:  Windows 11 Business 23H2

### PR DESCRIPTION
**Pull request by ErikEJ**
Also makes some tests run on machines in winter outside UTC time zones :-D

fixes #2209

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/MetadataUtilsSmi.cs [hunk 1]
```diff
@@ -244,6 +256,16 @@ Type udtType
                             extendedCode = ExtendedClrTypeCode.Char;
                         break;
                     case SqlDbType.Date:
+#if NET6_0_OR_GREATER
+                        if (value.GetType() == typeof(DateOnly))
+                            extendedCode = ExtendedClrTypeCode.DateOnly;
+                        else if (value.GetType() == typeof(DateTime))
```
Comment (ErikEJ) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/MetadataUtilsSmi.cs:262]: Wonder if these additional if branches are actually needed?

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/MetadataUtilsSmi.cs [hunk 2]
```diff
@@ -330,6 +352,14 @@ Type udtType
                             extendedCode = ExtendedClrTypeCode.Invalid;
                         }
                         break;
+#if NET6_0_OR_GREATER
+                    case SqlDbType.Time:
+                        if (value.GetType() == typeof(TimeOnly))
+                            extendedCode = ExtendedClrTypeCode.TimeOnly;
+                        else if (value.GetType() == typeof(TimeSpan))
```
Comment (ErikEJ) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/MetadataUtilsSmi.cs:359]: Wonder if this additional if branch is actually needed?

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs [hunk 3]
```diff
@@ -2899,6 +2907,10 @@ int length
 /*EnSDR*/{ _ ,  _ ,  _ , _ , _ , _  , _ , _ , _  , _  , _  , _  , _  , _ , _ , _  , _ , _  , _  , _ , _ , _  , _  , _  , _ , _  , _ , _ , _ ,  _ , X,  _  , _  , _  , _ , },/*IEnurerable<SqlDataRecord>*/
 /*TmSpn*/{ _ ,  _ ,  _ , _ , _ , _  , _ , _ , _  , _  , _  , _  , _  , _ , _ , _  , _ , _  , _  , _ , _ , _  , _  , _  , _ , _  , _ , _ , _ ,  _ , _,  _  , X  , _  , _ , },/*TimeSpan*/
 /*DTOst*/{ _ ,  _ ,  _ , _ , _ , _  , _ , _ , _  , _  , _  , _  , _  , _ , _ , _  , _ , _  , _  , _ , _ , _  , _  , _  , _ , _  , _ , _ , _ ,  _ , _,  _  , _  , _  , X , },/*DateTimeOffset*/
+#if NET6_0_OR_GREATER
```
Comment (ErikEJ) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs:2910]: I have NO idea if this is "correct" - copied from the row for DateTime/TimeSpan

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs [hunk 4]
```diff
@@ -2951,6 +2963,10 @@ int length
...
```
Comment (ErikEJ) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs:2966]: I have NO idea if this is "correct" - copied from the row for DateTime/TimeSpan

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlMetaDataTest.cs [hunk 5]
```diff
@@ -1019,7 +1019,7 @@ public void XmlConstructorWithNullObjectName_Throws()
             new object[] {SqlDbType.VarBinary, new byte[8001]},
             new object[] {SqlDbType.Time, new TimeSpan(0, 0, 1)},
             new object[] {SqlDbType.Time, new TimeSpan(TimeSpan.TicksPerDay - 1)},
-            new object[] {SqlDbType.DateTimeOffset, new DateTimeOffset(new DateTime(0))},
+            new object[] {SqlDbType.DateTimeOffset, new DateTimeOffset(new DateTime(0), TimeSpan.Zero)},
```
Comment (ErikEJ) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlMetaDataTest.cs:1022]: Make test pass during DTS

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs (hunk 3 above)
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs:2910]: Looks ok to me clr DateOnly to the various sql Date types and var. TimeOnly only to Timespan.
That entire thing is copied from a grid in the docs somewhere about type conversions but i've no idea if it's been updated to include dateonly and timeonly.

**Comment by ErikEJ** [comment]
@David-Engel @JRahnama Any chance to get this reviewed now?

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs (hunk 3 above)
Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs:2910]: I have no idea either... 👀 I'll trust that the tests are validating that this is correct.

**Review by David-Engel — APPROVED** [review]
I think this will end up in the release after the 5.2 GA.

**Comment by ErikEJ** [comment]
@David-Engel @JRahnama Is this still on the radar?

**Comment by ErikEJ** [comment]
@David-Engel @JRahnama Is this still on the radar?

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs (hunk 3 above)
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/ValueUtilsSmi.cs:2910]: It seems to be used to map the compatibility of SQL data types with extended CLR type codes. I found the following documentation the most relevant, but it's challenging to infer from there!
[SQL-CLR type mapping](https://learn.microsoft.com/en-us/dotnet/framework/data/adonet/sql/linq/sql-clr-type-mapping)

**Review by DavoudEshtehari — APPROVED** [review]

**Pull request merged**